dlib/cuda/cuda_dlib.cu

```diff
@@ -1377,12 +1377,25 @@ namespace dlib
 
         __global__ void _cuda_leaky_relu_gradient(float* out, const float* s, const float* gi, size_t n, const float alpha)
         {
-            for (auto i : grid_stride_range(0, n))
+            if (out == gi)
```

Comment:
A branch like this is relatively costly in CUDA. If you broke this up into two separate kernels it would likely run measurably faster.  The launching code can then contain the branch which happens once and CUDA never sees it.

Reply:
Thank you for that advice!
One quick question, I've updated the code to get:
```cpp
float* out = grad.device();
const float* gi = gradient_input.device();
```
before making the comparison and launching the kernel. Even if from the documentation of `tensor::to_device()`, it seems that if I call `.to_device()` twice in a row, nothing happens,, it seemed preferable to avoid such calls.